### PR TITLE
Added support for factory functions

### PR DESCRIPTION
**Pull request by slavafomin**
Hello!

This addresses #10.

I've implemented a basic support for factory functions, added tests and updated the documentation.

Could you please take a look and tell me what you think?

I hope the next step will be to introduce factory support for decorators.

### CODE REVIEW

File: src/Container.ts [hunk 1]
```diff
@@ -89,8 +92,16 @@ export class Container {
             params.unshift(null);
         }
 
-        // create a new instance of the requested object
-        const objectInstance = new (type.bind.apply(type, params))();
+        let objectInstance: any;
+
+        if (factory) {
+            // Calling factory function if specified to create an instance.
+            objectInstance = factory.apply(undefined, params);
```
Comment (pleerock) [inline src/Container.ts]: you probably can simply do `factory(params)` ?

File: test/unit/Container.spec.ts [hunk 2]
```diff
@@ -98,4 +113,14 @@ describe("Container", function() {
         Container.get(ExtraService).byeMessage.should.be.equal("buy world");
     });
 
+    it("should support class factories", function() {
+
+        const serviceFactory = new ServiceFactory("Hello Factory");
+
+        Container.registerService(undefined, FactoriedService, undefined, serviceFactory.create.bind(serviceFactory));
```
Comment (pleerock) [inline test/unit/Container.spec.ts]: You example (and especially using `bind` I find complex. Better if you did `(params) => serviceFactory.create(params)` in my opinion. Also using class not necessary here at all, you can simply check it this way:

```ts
Container.registerService(undefined, FactoriedService, undefined, () => new ServiceFactory("Hello factory"))
```

This would simplify understanding of this test by others.

File: test/unit/Container.spec.ts [hunk 3]
```diff
@@ -98,4 +113,14 @@ describe("Container", function() {
         Container.get(ExtraService).byeMessage.should.be.equal("buy world");
     });
 
+    it("should support class factories", function() {
```
Comment (pleerock) [inline test/unit/Container.spec.ts]: can you also add a test for the case when FactoriedService depend on another service to make sure `params` work as expected

File: src/Container.ts [hunk 4]
```diff
@@ -42,15 +42,16 @@ export class Container {
      *
      * @param name Service name. Optional
      * @param type Service class
-     * @param params Parameters to be sent to constructor on service initialization
+     * @param params Parameters to be sent to a class constructor or factory function on service initialization
+     * @param factory Factory function to be called on service initialization
      */
-    static registerService(name: string, type: Function, params?: any[]) {
-        this.registeredServices.push({ name: name, type: type, params: params });
+    static registerService(name: string, type: Function, params?: any[], factory?: Function) {
```
Comment (pleerock) [inline src/Container.ts]: probably we can rename this method to `register` and introduce an interface here to register service this way:

```ts
Container.register({
     name?: string,
     type: Function,
     params?: any[],
     factory?: (params?: any[]) => Function
});
```

**Comment by pleerock** [comment]
> I hope the next step will be to introduce factory support for decorators.

yes it can be:

```ts
@Service({ name?: string,  factory?: (params?: any[]) => Function  })
class MyService {

}
```

File: src/Container.ts (hunk 1 above)
Comment (slavafomin) [inline src/Container.ts]: If you would call `factory(params)` instead of `apply()` then all parameters would be passed as an array to the first argument of the factory function. I think it's undesired behavior? Shouldn't the elements of `params` array be mapped to the arguments of the factory function?

File: test/unit/Container.spec.ts (hunk 2 above)
Comment (slavafomin) [inline test/unit/Container.spec.ts]: I agree, it was just a use case I was interested in. I will simplify this test.

File: test/unit/Container.spec.ts (hunk 3 above)
Comment (slavafomin) [inline test/unit/Container.spec.ts]: @pleerock As I understand mocha runs each test serially, without clearing the intermediate state. It makes testing difficult and error-prone, because container is filling with different services with each individual test. I think we need to implement an approach when Container is cleared or re-created before each individual test. It will make it more straightforward and robust.

However, the Container class is static right now, so we can't just re-create it. Do you have ideas how it can be solved? Do we create a `clear()` method to reset the container or do we make a container a singleton instance with ability to re-create itself?

File: src/Container.ts (hunk 4 above)
Comment (slavafomin) [inline src/Container.ts]: Shouldn't the type of factory function be: `(...params: any[]) => Function` instead of `(params?: any[]) => Function`? Because, it could accept multiple arguments, not just a single array.

Comment (slavafomin) [inline src/Container.ts]: Also, the return type should be `Object` not `Function`, because factory function returns instances, not functions. So the type of the factory function should be: `(...params: any[]) => Object`. Do you agree?

**Comment by slavafomin** [comment]
@pleerock I've pushed some changes according to your comments. However, the last test is not working. I'm not sure what `params` do I need to pass to the `registerService()` in order to specify dependencies for the factory function. Could you help with that?

Also, if `registerService()` is a public API, shouldn't it use Reflect to get metadata of the factory function in order to resolve it's parameters automatically?

File: src/Container.ts (hunk 1 above)
Comment (pleerock) [inline src/Container.ts]: maybe you are right... then in this case you can do `factory(...params)`

Comment (pleerock) [inline src/Container.ts]: but Im not sure about this design. Problem with this design is if someday we will need to pass anything else (for example container instance, or some options) we can't do it

File: src/Container.ts (hunk 4 above)
Comment (pleerock) [inline src/Container.ts]: you are right, it should be `Object`. Regarding to `...params: any[]` I answered in previous comment.

File: test/unit/Container.spec.ts (hunk 3 above)
Comment (pleerock) [inline test/unit/Container.spec.ts]: > I think we need to implement an approach when Container is cleared or re-created before each individual test.

I think you are right. But the question shall it be a public `clear` method or shall we dynamically reset all properties this way: `(Container as any)["instances"] = (Container as any)["paramHandlers"] = ... = [];`

**Comment by pleerock** [comment]
If you want to move `(<any> Reflect).getMetadata("design:paramtypes", target)` to container then go ahead, I don't see reason why it wont work there. This probably will solve your issue with `params` and how to get `Engine` instance in tests.

File: src/Container.ts (hunk 1 above)
Comment (slavafomin) [inline src/Container.ts]: @pleerock Why do **we** need to pass something to the factory function? The end-user should decide what should be passed to it.

File: test/unit/Container.spec.ts (hunk 3 above)
Comment (slavafomin) [inline test/unit/Container.spec.ts]: Actually, I'm starting to think that better design would be to allow end-user to have multiple separate containers, each with it's own list of registered services and created instances. This will make the container more flexible and it will solve the problem mentioned above for us, i.e. we will be able to just create another container inside of each individual test.

The API could look like this:

```ts
const container1 = Container.create();
container1.registerService(/* Foo */);
container1.registerService(/* Bar */);

const container2 = Container.create();
container2.registerService(/* Baz */);

const bar = container1.get(Bar);
bar.doBar();

const baz = container2.get(Baz);
baz.doBaz();

const foo = container2.get(Foo); // Error: Foo service is not registered

```

I think Symfony is a great example of a quality code, we could try to make a similar API of the Symfony container: http://symfony.com/doc/current/components/dependency_injection.html. As you can see, the container is a normal class, not a static class there: https://github.com/symfony/dependency-injection/blob/master/Container.php.

What do you think about such design? Of course we are talking about breaking changes here, but according to versioning this module is not yet stable ( < 1.0.0).

File: src/Container.ts (hunk 1 above)
Comment (pleerock) [inline src/Container.ts]: there may be cases in the future when it *may* be needed

File: test/unit/Container.spec.ts (hunk 3 above)
Comment (pleerock) [inline test/unit/Container.spec.ts]: the problem with container instance (unlike singleton or static service) that we will need to tell  to DI implicitly what services are defined. Because decorators can't have instance of container where they can store data about decorated classes.

File: src/Container.ts (hunk 1 above)
Comment (slavafomin) [inline src/Container.ts]: Could you please provide an example or a an edge case, usage scenario? My imagination is failing me here.

File: test/unit/Container.spec.ts (hunk 3 above)
Comment (slavafomin) [inline test/unit/Container.spec.ts]: Yes, that makes sense. What if we introduce a namespaces concept or a named containers then?

```ts
// This is the container without a name, it's a default one:
const container1 = Container.create();

// This is additional container which goes under the "second-container" name:
const container2 = Container.create("second-container");

// This service belongs to the default container.
@Service()
class Foo {
}

// However, this service belongs to the second container.
@Service("bar", "second-container")
class Bar {
}

```

Or more meaningful API:

```ts
@Service({
  name: "bar",
  container: "second-container"
})
class Bar {
}
```

In order to replace a container:

```ts
const container1 = Container.create();
const container2 = Container.create("second-container");

// Use one of these interchangeably:
Container.replace("default", "second-container");
Container.replace("default", container2);
Container.replace(container1, container2);
// "second-container" is now used as a default, the default one is effectively destroyed

// These two will replace the default container with a new empty one.
const newDefault = Container.replace("default");
const newDefault = Container.replace();
```

This looks like a viable solution, but it also looks a bit complicated. What do you think?

**Comment by slavafomin** [comment]
@pleerock I've made another change. Specifically, I've introduced another decorator for factory methods and added separate facility for factories registration. The test is now working correctly, and factory arguments are also properly injected. Could you please take a look and let me know if this implementation makes sense? I'm still looking for other solutions though.

File: src/Container.ts (hunk 1 above)
Comment (pleerock) [inline src/Container.ts]: I cant answer your question because I don't have examples. These cases come unexpectedly. There are real-life situations when you are using ...params and some day you are realizing you need to pass some option-x and you can't do it at all because of design you choose. Okay, lets go with ...params syntax

File: test/unit/Container.spec.ts (hunk 3 above)
Comment (pleerock) [inline test/unit/Container.spec.ts]: Good solution I think. But I didn't get point of container `replace`. Looks weird and I dont see use cases of this

File: src/Container.ts [hunk 5]
```diff
@@ -96,18 +106,29 @@ export class Container {
...
         }
 
-        let objectInstance: any;
-
         if (factory) {
-            // Calling factory function if specified to create an instance.
-            objectInstance = factory.apply(undefined, params);
+
+            // Calling factory method to create an instance.
+            objectInstance = factory.factoryFunction.apply(undefined, params);
```
Comment (pleerock) [inline src/Container.ts]: it can be `factory.factoryFunction(...params)` instead. Btw is there a better name for `factory.factoryFunction` ?

File: test/unit/Container.spec.ts [hunk 6]
```diff
@@ -128,13 +112,13 @@ describe("Container", function() {
             }
         }
 
-        function createCar (engine: Engine) {
-            return new Car(engine);
+        class CarFactory {
+            @Factory()
+            public static createCar (engine: Engine): Car {
```
Comment (pleerock) [inline test/unit/Container.spec.ts]: will reflect metadata work if you remove `: Car` from the end of this line? I guess no. And if no, then this method is not good - its hidden and unclear for a user to put return type explicitly when he have a type inference.

Comment (pleerock) [inline test/unit/Container.spec.ts]: alternative can be to use `@Factory(Car)` syntax.

Comment (pleerock) [inline test/unit/Container.spec.ts]: also is it okay that we can defined such "factory" anywhere? I don't think so. Just imagine you have a big codebase and you have lets say `Car` service and somewhere deep in the app I can create a `EverythingFactory` and put there logic of car creation... I think it should be explicit and we should put `@Service({ factory: carFactory }) export class Car...` to make it really clear and make flow explicit. If you don't like that your function will be inside the decorator - then don't worry - it may not be there. Users just need to create `export function carFactory(){ ... }` in a separate file and use `@Service({ factory: carFactory })` where `carFactory` is imported from that file.

File: test/unit/Container.spec.ts (hunk 3 above)
Comment (slavafomin) [inline test/unit/Container.spec.ts]: OK, fair enough. I was thinking to use this replace functionality in tests where we can replace active "dity" container with a new "pristine" one. Right now, it looks like a better idea would be to implement "container.clear()" for this usage scenario.

Comment (slavafomin) [inline test/unit/Container.spec.ts]: @pleerock I'm going to extract these two features into separate issue, because it's outside of this PR.

Comment (slavafomin) [inline test/unit/Container.spec.ts]: https://github.com/pleerock/typedi/issues/12

Comment (slavafomin) [inline test/unit/Container.spec.ts]: [Make containers named and instantiable](https://github.com/pleerock/typedi/issues/13)

File: src/Container.ts (hunk 5 above)
Comment (slavafomin) [inline src/Container.ts]: > `factory.factoryFunction(...params)`

Cool! Didn't know that. Thanks )

> Btw is there a better name for factory.factoryFunction ?

It comes from here: https://github.com/slavafomin/typedi/blob/master/src/Container.ts#L69

I guess we can call it just a "function", but "factoryFunction" looks more clear, I think. Maybe you have a better idea?

File: test/unit/Container.spec.ts (hunk 6 above)
Comment (slavafomin) [inline test/unit/Container.spec.ts]: Actually, I think that concepts of `@Service` and `@Factory` decorators are very similar. The difference is that service is created by calling the constructor of the class directly and factory is used to instantiate another class. Both of them have some "magic" in it. The entire concept of autowiring is about implicit configuration. So, I believe this could be acceptable. In order to mitigate such conflicts we can explicitly check if factory is already registered for a specified service and throw an exception to make such conflicts apparent.

However, I agree that `@Service({ factory: carFactory })` syntax is more expressive and I like it better. Maybe we could have both of them?

The more serious question is I can't really define how the factory itself should look like. Could it be a simple function? Or should it be a static class with static function? Of should it be a normal class, which itself must be created before it's method is called in order to create a service? Probably, we should support all these scenarios? In such case `@Factory` decorator is not an option, because it can't be added to a simple function due to language limitations.

Finally, how do we support all those use cases? I will try to summarize them:

```ts
class Car {
  constructor (private engine: Engine) {
  }
}
```

**1. The factory is just a simple standalone function.**

```ts
@Service({ factory: createCar })
class Car {
  constructor (private engine: Engine) {
  }
}

function createCar (Engine: engine): Car {
  return new Car(engine);
}
```

**2. The factory is a static function of a class.**

```ts
@Service({ factory: CarFactory.createCar })
class Car {
  constructor (private engine: Engine) {
  }
}

class CarFactory {
  public static createCar (Engine: engine): Car {
    return new Car(engine);
  }
}
```

**3. The factory is a normal function of a class instance.**

```ts
@Service({ factory: ... })
class Car {
  constructor (private engine: Engine) {
  }
}

class CarFactory {
  construct (private logger: Logger) {
  }
  public createCar (Engine: engine): Car {
    this.logger.debug("Creating a new car!");
    return new Car(engine);
  }
}
```

---

The first two scenarios looks almost the same, the function is called directly.

However, the third case is interesting, because in order to call a factory function we need to create an instance of a factory class first. So the factory class could be a service in itself. The question is how should we specify both the class of the factory and the method?

`@Service({ factory: [CarFactory, CarFactory.prototype.createCar] }) class Car {}` or

`@Service({ factory: [CarFactory, "createCar"] }) class Car {}` ?

The first variant looks longer, but more specific and could be parsed and understood by IDE.

What do you think about all this?

Comment (pleerock) [inline test/unit/Container.spec.ts]: `@Factory` decorator should be removed to remove "double-behaviour" and remove cases when service is initialized without knowing where how and who affected service creation.

For your second question, function in factory is enough. I think there is no need to complicate things with such complex use cases.

File: src/Container.ts (hunk 5 above)
Comment (pleerock) [inline src/Container.ts]: same opinion as yours. lets left it `factoryFunction` then.

File: test/unit/Container.spec.ts (hunk 6 above)
Comment (slavafomin) [inline test/unit/Container.spec.ts]: Okay, I'll remove `@Factory`. Considering the other question — the problem is, we can't decide for the user. She could have a preexisting system with factories defined in one of these three ways (or in mixed ways). The container should be able to handle all cases the same. What do you think about the third case? How should we pass the factory class as well as a factory function to the container?

Actually, there is another case — when factory is already registered with the container using named service. We should also allow a way to reference the factory using a service name. The API could look like this:

```ts
@Service({ factory: ["car-factory", CarFactory.prototype.createCar] }) class Car {}
                                         ^--- Name of the service registered with the container
```

Comment (pleerock) [inline test/unit/Container.spec.ts]: > we can't decide for the user.

we can't provide everything user wants as well. Lets skip third case for now. If there be many people requesting this feature we will start to think how to implement it the best way. But I doubt many people need such complex functionality.

**Comment by slavafomin** [comment]
@pleerock I've pushed some changes as we discussed:

1. I've introduced a common interface for service descriptor.
2. I've removed separate facilities for factories.

However, I now have the problem which I hope you will be able to help me to solve:

How do I get parameters of the factory function from inside the `@Service` decorator?
https://github.com/slavafomin/typedi/blob/master/src/decorators.ts#L17
Is it even possible to do so? Or parameters could only be retrieved from the decorator of the method itself? If so, we can't support simple functions as factories, only class methods (please correct me if I'm wrong). And we will still need to use a separate decorator for the factory method.

File: README.md [hunk 7]
```diff
@@ -212,6 +212,40 @@ let coffeeMaker = Container.get<CoffeeMaker>("coffee.maker");
...
+  }
+}
+
+function createService () {
+  return new Service("Built with factory!");
+}
+
+// Registering service with container.
+// You can also pass additional arguments for your factory function.
+Container.registerService("my-service", Service, [], createService);
```
Comment (pleerock) [inline README.md]: you forgot to update docs

File: src/Container.ts [hunk 8]
```diff
@@ -1,5 +1,14 @@
+
 import {ParamHandler, PropertyHandler} from "./Handlers";
 
+
+export interface ServiceDescriptor {
```
Comment (pleerock) [inline src/Container.ts]: please extract this into a separate file

File: src/Container.ts [hunk 9]
```diff
@@ -40,38 +49,42 @@ export class Container {
...
      */
     static get<T>(type: ConstructorFunction<T>, params?: any[]): T;
     static get<T>(name: string, params?: any[]): T;
     static get<T>(typeOrName: ConstructorFunction<T>|string, params?: any[]): T {
 
         // normalize parameters
-        let type: Function, name: string;
+        let type: Function;
+        let name: string;
+        let factory: Function;
```
Comment (pleerock) [inline src/Container.ts]: I think one line is better then tree in this case

File: src/Container.ts [hunk 10]
```diff
@@ -40,38 +49,42 @@ export class Container {
...
+        if (serviceDescriptor) {
+            if (!type) {
+                type = serviceDescriptor.type;
+            }
+            if (!params) {
+                params = serviceDescriptor.params;
+            }
+            if (serviceDescriptor.factory) {
+                factory = serviceDescriptor.factory;
+            }
```
Comment (pleerock) [inline src/Container.ts]: please avoid putting `{` and `}` to a short single-line `if`

File: src/Container.ts [hunk 11]
```diff
@@ -83,14 +96,27 @@ export class Container {
...
             params = this.initializeParams(type, params);
-            params.unshift(null);
+            if (!factory) {
+                params.unshift(null);
+            }
+        }
+
+        if (factory) {
+
+            // Calling factory function to create an instance.
```
Comment (pleerock) [inline src/Container.ts]: please start single-line comment block with a small-case latter

File: src/Container.ts [hunk 12]
```diff
@@ -178,23 +204,21 @@ export class Container {
...
+    private static findRegisteredService(name: string, type: Function): ServiceDescriptor {
         if (name) {
             return this.findRegisteredServiceByName(name);
         } else if (type) {
             return this.findRegisteredServiceByType(type);
         }
     }
 
-    private static findRegisteredServiceByType(type: Function) {
+    private static findRegisteredServiceByType(type: Function): ServiceDescriptor {
```
Comment (pleerock) [inline src/Container.ts]: 👍

File: src/decorators.ts [hunk 13]
```diff
@@ -1,15 +1,31 @@
...
 
 /**
  * Marks class as a service that can be injected using container.
  *
- * @param name Optional service name can be specified. If service name is specified then this service can only be
- *              retrieved by a service name. If no service name is specified then service can be retrieved by its type
+ * @param descriptor
  */
-export function Service(name?: string) {
+export function Service(descriptor: ServiceDescriptor = {}) {
```
Comment (pleerock) [inline src/decorators.ts]: `descriptor: ServiceDescriptor = {}` syntax does not look good. Better to make it `descriptor?`. Extra check for descriptor in the code worth it.

File: src/decorators.ts [hunk 14]
```diff
@@ -1,15 +1,31 @@
...
+                descriptor.params = (<any> Reflect).getMetadata("design:paramtypes", target);
+            }
+        }
+        Container.registerService(descriptor);
+    };
+}
+
+export function Factory() {
+    return function (target: any, propertyKey: string, descriptor: PropertyDescriptor) {
+      // @todo: maybe we still need it...
```
Comment (pleerock) [inline src/decorators.ts]: not in this life 😈

File: test/unit/decorators.spec.ts [hunk 15]
```diff
@@ -19,7 +19,7 @@ describe("Service Decorator", function() {
         @Service()
         class TestService {
         }
-        @Service("super.service")
+        @Service({ name: "super.service" })
```
Comment (pleerock) [inline test/unit/decorators.spec.ts]: no no no no ! We need to add two overloads of `@Service` decorator:

```ts
@Service("super.service")
@Service({ ... })
```

**Comment by pleerock** [comment]
> How do I get parameters of the factory function from inside the @Service decorator?
Is it even possible to do so? Or parameters could only be retrieved from the decorator of the method itself? 

It should work we just need to move Reflect.getMetadata.... code to the container I think. Need to check it. Try it and if it won't work I'll take a look by myself.

File: README.md (hunk 7 above)
Comment (slavafomin) [inline README.md]: Yeah, we will need to make it work first. Then I will be glad to update the docs.

File: src/Container.ts (hunk 8 above)
Comment (slavafomin) [inline src/Container.ts]: I was thinking the same.

File: src/Container.ts (hunk 9 above)
Comment (slavafomin) [inline src/Container.ts]: One-liners is always harder to read and I believe it's a part of most of the style guides I've seen.

File: src/Container.ts (hunk 10 above)
Comment (slavafomin) [inline src/Container.ts]: I think, this is also considered an agreed-upon style guide in the industry and is also easier to read.

**Comment by slavafomin** [comment]
@pleerock I've already tried and I can't get metadata from the factory function this way. Could you please take a look?

Considering the style guides I would suggest to add your requirements to the linting configuration, so other developers will be able to see errors inside of their IDEs. Also, are you basing this on your own preferences, or are you targeting some community-accepted/popular standard?

File: src/decorators.ts (hunk 13 above)
Comment (slavafomin) [inline src/decorators.ts]: This will make the code more complex and harder to read. I'm not really sure how you want it to be.

**Comment by slavafomin** [comment]
@pleerock Could you please also look into overloading of the `@Service` decorator to support both name string and descriptor object as an input parameter? I'm new to TS and I'm not sure how to correctly implement it.

File: src/Container.ts (hunk 9 above)
Comment (pleerock) [inline src/Container.ts]: its just definition part noone really look on it

File: src/Container.ts (hunk 10 above)
Comment (pleerock) [inline src/Container.ts]: `in the industry` where? in c++ or php?). Styles guides are different, some of them popular some of them less popular. In php styleguide with always-{} is more popular. In javascript I don't think its more popular. I always was against {} in simple single line statements because it makes code uglier and more complicated. Simply need to know when you to put {} in a single line statements and when not, and format with above and below newlines properly.

File: src/decorators.ts (hunk 13 above)
Comment (pleerock) [inline src/decorators.ts]: The most important always is public api. Which one do you think is better clear and straightforward for user `@Service(descriptor: ServiceDescriptor = {}) ` or `Service(descriptor?: ServiceDescriptor)`.  Implementation is just one extra check.

Also, is `ServiceDescriptor` isn't best name here. Better to do `@Service(options?: ServiceOptions)`. And better to left this object dummy and create a new object `ServiceMetadata` and copy all options into it. I'm following this pattern in all libs. If you are confused and don't get the idea just left it to me, I'll refactor once I merge these changes.

**Comment by pleerock** [comment]
Style guide is based on my rich experience. JavaScript community is large and multiopinion and you can find popular style guide, unlike php where community is much smaller and less active and there is mostly one style guide dominating. But I'm not sure if there is strong opinion for the things we were talking about because they are case-base. Its good to configure those options in lslint if they are supported.

**Comment by pleerock** [comment]
> @pleerock Could you please also look into overloading of the @Service decorator to support both name string and descriptor object as an input parameter? I'm new to TS and I'm not sure how to correctly implement it.

sure. please finish everything you left/need/want to do and I'll finish rest.

File: src/Container.ts (hunk 9 above)
Comment (slavafomin) [inline src/Container.ts]: The human eye works in "F" pattern (there are studies), i.e. it scans lines vertically first, then it scans the interesting lines horizontally. So it's better to keep important stuff closer to the start of the line and to split it into multiple lines. For me it's much easier to find `name` variable in this case when it's defined on it's own line. Also, when each variable has it's own definition, we can quickly change it's type from `const` to `let` for example, otherwise we will have to split them beforehand.

File: src/Container.ts (hunk 10 above)
Comment (slavafomin) [inline src/Container.ts]: Actually it makes the code more standardized (you have only one way to define code block instead of two) and it makes it easier to modify, because you can quickly add something to the existing code block: for example you can add an ugly `console.log()` statement to debug something. Otherwise you will have to first convert single line to a code block and then add a statement to it. And if that statement was temporary you will have to convert it back to a single line after deleting it. This is a lot of extraneous work. I was using single-line approach in PHP earlier in my career, but then switched to always using code blocks and I very rarely see code where single line statements are used instead of code blocks nowadays. And it's not making code harder to read (objectively), it's just a matter of preference and habit, I think.

File: src/decorators.ts (hunk 13 above)
Comment (slavafomin) [inline src/decorators.ts]: @pleerock Thank you, please do this. I will try to learn something from your implementation then.

**Comment by slavafomin** [comment]
@pleerock I've reverted name based decorators in tests to use single string and also updated the README.

The tests are not working now. What needs to be done:

1). Introduce overloading for `@Service` decorator to support both single string and object calls, this should fix most of the tests.

2). Make service factory decorators work, we need to resolve factory function arguments somehow.

3). Make `Container.registerService()` work with factories, I don't know how to manually pass the arguments in test.

4). This is closely related to "2" and "3" above: I think it would be better to move arguments reflection code to the central service registration facilit, like we discussed (it will allow to register services both via decorators and via manual call to the container, no need to manually specify arguments like in "3").

I will be glad to assist further if you need my help. Thank you!

File: src/Container.ts (hunk 9 above)
Comment (pleerock) [inline src/Container.ts]: Right, you can quickly change almost as quick as extracting it into a new line. so never use `quickly change` as argument for a small change cases. `So it's better to keep important stuff ` - right thats what my point is - this part is not important, its just regular variable declaration and can be in a single line to not distract from other important lines of code ;)

File: src/Container.ts (hunk 10 above)
Comment (pleerock) [inline src/Container.ts]: there is no need to make such things standartized because sometimes its really better to put and sometimes its not. Better to choose best cases instead of following dummy (in this case) standartization. 

No need to show example where you can easily change "change line" - you can do it almost as well easily as putting {} and add extra line. Yeah I know it may be annoying sometimes but code beauty worth it, and its totally not a "lot of extraneous work", just few seconds change. 

I also changed my opinion multiple times during my career regarding to those braces. First without {} then with. Your arguments are standard PHP-guy arguments (I saw lot of them arguing this :D), I also used these arguments in past, but nowadays and already 4-5 years stable I prefer put them depend on case where necessary. This makes the code best and clear. Yeah I know code style disputes never stop

**Comment by pleerock** [comment]
Okay, got you, thanks for so many things done. Nevermind on style guide notices, I can change it by myself. I'll work on it tomorrow.
